File: 0881-boats-to-save-people/0881-boats-to-save-people.js
/**
 * @param {number[]} people
 * @param {number} limit
 * @return {number}
 */
var numRescueBoats = function(people, limit) {
    people.sort((a,b) => a - b); // complexity n
    let lightest = 0;
    let heaviest = people.length - 1;
    let boats = 0;
    for (let i in people) {
        if (lightest === heaviest) {
            boats++;
            break;
        }
        if (lightest > heaviest) {
            break;
        }
        if (people[lightest] + people[heaviest] <= limit) {
            boats++;
            lightest++;
            heaviest--;
        } else {
            boats++;
            heaviest--;
        }
    }
    return boats;
};

File: 0881-boats-to-save-people/0881-boats-to-save-people.py
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()
        light = 0
        heavy = len(people) - 1
        boats = 0
        
        while light <= heavy:
            if (light == heavy):
                boats += 1
                break

            if people[light] + people[heavy] <= limit:
                light += 1

            heavy -= 1
            boats += 1
        
        return boats